Readonly classes cannot be combined with `#[AllowDynamicProperties]`

staabm opened the pull request:
closes https://github.com/phpstan/phpstan/issues/12281

staabm committed:
Readonly classes cannot be combined with `#[AllowDynamicProperties]`

staabm commented on tests/PHPStan/Rules/Classes/data/bug-12281.php:
```diff
@@ -0,0 +1,10 @@
+#[\AllowDynamicProperties]
+readonly class BlogData { /* … */ }
+
+/** @readonly  */
+#[\AllowDynamicProperties]
+class BlogDataPhpdoc { /* … */ }
```
not sure we want the same error for phpdoc readonly, as we have for native readonly...?

looking thru the source we usually only do the "readonly-by-phpdoc" dance for properties, not for classes

ondrejmirtes replied:
I wouldn't do that for PHPDoc `@readonly`.

ondrejmirtes commented on src/Rules/Classes/ClassAttributesRule.php:
```diff
@@ -28,12 +31,23 @@ public function processNode(Node $node, Scope $scope): array
+		$classReflection = $scope->getClassReflection();
+		if ($classReflection !== null && $classReflection->isReadOnly()) {
+			if (count($classReflection->getNativeReflection()->getAttributes('AllowDynamicProperties')) > 0) {
+				$errors[] = RuleErrorBuilder::message(sprintf('Cannot apply #[AllowDynamicProperties] to readonly class %s.', $classReflection->getName()))
```
We don't use this error message format anywhere. I'd follow some error from AttributesCheck. Example:

> Attribute class AllowDynamicProperties cannot be used with readonly classes.

ondrejmirtes commented on src/Rules/Classes/ClassAttributesRule.php:
```diff
@@ -28,12 +31,23 @@ public function processNode(Node $node, Scope $scope): array
+			if (count($classReflection->getNativeReflection()->getAttributes('AllowDynamicProperties')) > 0) {
+				$errors[] = RuleErrorBuilder::message(sprintf('Cannot apply #[AllowDynamicProperties] to readonly class %s.', $classReflection->getName()))
+					->identifier('class.allowDynamicPropertiesReadonly')
+					->build();
```
Should be non-ignorable

ondrejmirtes reviewed (changes requested):
We also need to error when this attribute is used above interfaces. Also please check what happens above enums and traits (I'm not sure).

staabm commented:
> We also need to error when this attribute is used above interfaces. Also please check what happens above enums and traits (I'm not sure).

I have tested this cases and linked the 3v4l.org examples in https://github.com/phpstan/phpstan/issues/12281#issue-2744490661

TL;DR:
- enums not allowed
- interfaces not allowed
- traits not allowed

what do you think about adding a separate `AllowDynamicPropertiesAttributeRule` which targets `ClassLike` instead of adding this error to the existing rule?
(atm we don't have a rule which covers attributes on traits)

ondrejmirtes commented on src/Rules/Classes/ClassAttributesRule.php:
```diff
@@ -28,12 +31,34 @@ public function processNode(Node $node, Scope $scope): array
 			'class',
 		);
+
+		$classReflection = $scope->getClassReflection();
```
You can ask `$node` for the class reflection, it's not nullable on that.

ondrejmirtes commented on src/Rules/Classes/ClassAttributesRule.php:
```diff
@@ -28,12 +31,34 @@ public function processNode(Node $node, Scope $scope): array
+				$message = 'Attribute class AllowDynamicProperties cannot be used with enums.';
+			}
+			if ($classReflection->isInterface()) {
+				$message = 'Attribute class AllowDynamicProperties cannot be used with interface.';
```
This could by DRYed and cleaned up with sprintf.

ondrejmirtes commented on src/Rules/Classes/ClassAttributesRule.php:
```diff
@@ -28,12 +31,34 @@ public function processNode(Node $node, Scope $scope): array
+
+			if (count($classReflection->getNativeReflection()->getAttributes('AllowDynamicProperties')) > 0) {
+				$errors[] = RuleErrorBuilder::message($message)
+					->identifier('class.allowDynamicPropertiesReadonly')
```
Needs better identifier for enums and interfaces.

ondrejmirtes reviewed (changes requested):
Also:

1) Please do new TraitAttributesRule similar to ClassAttributesRule that will call AttributesCheck. We need the rule to hook onto Trait_ and fetch the reflection ourselves from ReflectionProvider.
2) Add AllowDynamicProperties check there too.

ondrejmirtes commented on tests/PHPStan/Rules/Classes/data/bug-12281.php:
```diff
@@ -0,0 +1,19 @@
+interface BlogDataInterface { /* … */ }
+
+#[\AllowDynamicProperties]
+trait BlogDataTrait { /* … */ }
```
This test is irrelevant for this rule so move it elsewhere.

ondrejmirtes commented on src/Rules/Classes/ClassAttributesRule.php:
```diff
@@ -28,12 +31,39 @@ public function processNode(Node $node, Scope $scope): array
+			$identifier = 'class.allowDynamicPropertiesReadonly';
+			if ($classReflection->isEnum()) {
+				$typeName = 'enum';
+				$identifier = 'class.allowDynamicPropertiesEnum';
```
enum.allowDynamicProperties

ondrejmirtes commented on src/Rules/Classes/ClassAttributesRule.php:
```diff
@@ -28,12 +31,39 @@ public function processNode(Node $node, Scope $scope): array
+			}
+			if ($classReflection->isInterface()) {
+				$typeName = 'interface';
+				$identifier = 'class.allowDynamicPropertiesInterface';
```
interface.allowDynamicProperties

ondrejmirtes commented on src/Rules/Traits/TraitAttributesRule.php:
```diff
@@ -0,0 +1,61 @@
+		$errors = $this->attributesCheck->check(
+			$scope,
+			$node->attrGroups,
+			Attribute::TARGET_ALL, // there is no dedicated trait target
```
TARGET_CLASS should be used https://3v4l.org/DJPUf#v8.4.2

ondrejmirtes commented on src/Rules/Traits/TraitAttributesRule.php:
```diff
@@ -0,0 +1,61 @@
+		}
+
+		$errors = $this->attributesCheck->check(
+			$scope,
```
We need to do something with $scope here. Please recreate this problem https://github.com/phpstan/phpstan-src/commit/3447391001f7a5c2bfb77f66c2d0e157242c1dae but with a trait.

We probably need to do `$scope->enterTrait()` here before calling the check. Just do `if !$scope instanceof MutatingScope then throw ShouldNotHappen`.

staabm replied:
I have adjusted the PR. I think I have reproduced the problem you are after.

adding `$scope->enterTrait($classReflection)` didn't work and throws:
```
2) PHPStan\Rules\Traits\TraitAttributesRuleTest::testBug12011
PHPStan\ShouldNotHappenException: Internal error.

C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\ScopeContext.php:44
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\MutatingScope.php:2935
C:\dvl\Workspace\phpstan-src-staabm\src\Rules\Traits\TraitAttributesRule.php:49
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\FileAnalyser.php:111
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\NodeScopeResolver.php:482
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\NodeScopeResolver.php:354
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\NodeScopeResolver.php:842
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\NodeScopeResolver.php:307
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\FileAnalyser.php:212
C:\dvl\Workspace\phpstan-src-staabm\src\Analyser\Analyser.php:75
C:\dvl\Workspace\phpstan-src-staabm\src\Testing\RuleTestCase.php:177
C:\dvl\Workspace\phpstan-src-staabm\src\Testing\RuleTestCase.php:138
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Rules\Traits\TraitAttributesRuleTest.php:78
```

at this point I have no idea what todo. could you fix the remaining thing?

staabm replied:
let me fix the build real quick

ondrejmirtes replied:
Push the code that's crashing like that please, I will take a look

staabm replied:
here you are

ondrejmirtes replied:
Oh right, we need to be in a class first before we can enter a trait.

I'd try two things: try to put the trait reflection into enterClass (but a different ShouldNotHappen might be thrown, not sure).

If that fails, we need to come up with some fake class to put into enterClass before we can enterTrait. It might be okay to simulate some anonymous Class_ node to use.

If it's too much work, don't worry and I'll merge the original code that doesn't use MutatingScope at all.

staabm replied:
I have class-faking without a thrown exception working. the test-case is still failling though

Will have another look after sleeping over it

ondrejmirtes commented:
Please keep this on 2.0.x. We don't need to be adding new features to 1.12.x anymore.

staabm commented on src/Rules/Traits/TraitAttributesRule.php:
```diff
@@ -0,0 +1,61 @@
+		$classReflection = $this->reflectionProvider->getClass($traitName->toString());
+		if (count($classReflection->getNativeReflection()->getAttributes('AllowDynamicProperties')) > 0) {
+			$errors[] = RuleErrorBuilder::message('Attribute class AllowDynamicProperties cannot be used with trait.')
+				->identifier('class.allowDynamicPropertiesTrait')
```
maybe?

```suggestion
				->identifier('trait.allowDynamicProperties')
```

ondrejmirtes committed:
Update TraitAttributesRule.php

Co-authored-by: Markus Staab <markus.staab@redaxo.de>

staabm committed:
work in progress.. crashing atm

staabm committed:
fix class fake

ondrejmirtes commented:
I changed my mind, I'm going to adjust it myself, thanks!

ondrejmirtes committed:
Use InTraitNode instead

Some attribute arguments actually depend on using class instead of the trait

So we have to analyse the attributes in context of each class that uses the trait

ondrejmirtes committed:
Fix

ondrejmirtes commented:
See my commit and reasoning in description: https://github.com/phpstan/phpstan-src/pull/3738/commits/b1cebc2e1f5133b1bd68ebb4401947c5f4d0e324

ondrejmirtes commented:
Please stop force-pushing here, I said I'm taking over. Thanks.

staabm commented:
> Please stop force-pushing here, I said I'm taking over. Thanks.

oh sorry, did not see your comments. I got it working just a moment ago ;)

ondrejmirtes committed:
This file can still be linted

ondrejmirtes commented:
Thank you!

staabm commented:
thanks
